Buy/Graph.Client.swift

```diff
@@ -151,56 +150,41 @@ extension Graph {
-                DispatchQueue.main.async {
-                    
-                    if var retryHandler = retryHandler, retryHandler.canRetry, retryHandler.condition(response, error) == true {
-                        
-                        /* ---------------------------------
-                         ** A retry handler was provided and
-                         ** the condition evaluated to true,
-                         ** we have to retry the request.
-                         */
-                        retryHandler.repeatCount += 1
-                        
-                        let retryTask = self.graphRequestTask(query: query, retryHandler: retryHandler, completionHandler: completionHandler)
-                        task.setTask(retryTask.task)
-                        
-                        DispatchQueue.main.asyncAfter(deadline: .now() + retryHandler.interval) {
-                            if task.task.state == .suspended {
-                                task.resume()
-                            }
-                        }
-                        
-                    } else {
-                        completionHandler(response, error)
-                    }
-                }
-            }
-            
-            task = Task(representing: dataTask)
-            return task
+        private func graphRequestTask<Q: GraphQL.AbstractQuery, R: GraphQL.AbstractResponse>(query: Q, cachePolicy: CachePolicy, retryHandler: RetryHandler<R>? = nil, completionHandler: @escaping (R?, QueryError?) -> Void) -> Task {
+            
+            let request = self.graphRequestFor(query: query)
+            return InternalTask<R>(
+                session:      self.session,
+                cache:        self.cache,
+                request:      request,
+                cachePolicy:  cachePolicy,
+                retryHandler: retryHandler,
+                completion:   completionHandler
+            )
         }
         
         func graphRequestFor(query: GraphQL.AbstractQuery) -> URLRequest {
-            var request = URLRequest(url: self.apiURL)
+            var request     = URLRequest(url: self.apiURL)
+            let requestData = String(describing: query).data(using: .utf8)!
             
-            request.httpMethod = "POST"
-            request.httpBody = String(describing: query).data(using: .utf8)
+            request.httpMethod              = "POST"
+            request.httpBody                = requestData
             request.httpShouldHandleCookies = false
-            request.setValue("application/json", forHTTPHeaderField: Header.accept)
+            
+            request.setValue("application/json",    forHTTPHeaderField: Header.accept)
             request.setValue("application/graphql", forHTTPHeaderField: Header.contentType)
+            request.setValue(requestData.md5,       forHTTPHeaderField: Header.queryTag)
```

Comment:
It's not strictly necessary but it makes subsequent `hash` calculations much faster - `O(1)` vs `O(n)` where `n` is the size of the HTTP body in bytes. It's also very nice to have for debugging purposes both on client and server to be able to distinguish identical queries. It doesn't hurt anything to have this header in place.